### ISSUE TITLE
training stage: Expected predict_data <= 1, but received predict_data:-nan > 1:1. The predict data must less or equal 1

### ISSUE THREAD

**Issue by brandonsj** [PaddlePaddle/Paddle#15816]
PaddlePaddle version fluid 1.2
CPU on paddlecloud

meet fatal error during model training
paddle.fluid.core.EnforceNotMet: Enforce failed. Expected predict_data <= 1, but received predict_data:-nan > 1:1.
The predict data must less or equal 1. at [/script/paddle/paddle/fluid/operators/metrics/auc_op.h:80]

**Comment by brandonsj** [comment]
train.py
from __future__ import print_function
import sys
import pickle
import paddle
import paddle.fluid as fluid
import paddle.v2
import reader
import os
import math
import numpy as np
import time
import hybrid_dssm as dssm

USE_GPU = False
BATCH_SIZE = 256
PASS_NUM = 1
cluster_train_dir = "./train_data/"
cluster_test_dir = "./test_data/"


def train(use_cuda, save_dirname, is_local):

    feed_order = [
        'a_item_type','a_category_id','a_cluster_id','a_title_segs','a_tags',
        'b_item_type','b_category_id','b_cluster_id','b_title_segs','b_tags','label'
    ]

    feature_dic_shape = {}
    with open('./thirdparty/feature_dic.pkl', 'rb') as out:
        feature_dic = pickle.load(out)
    #for item in feature_dic:
    #    print(item)
    #    print(len(feature_dic[item]))

    for i in range(0, 10):
        feature_dic_shape[feed_order[i]] = len(feature_dic[feed_order[i]])
        print(feed_order[i])

    avg_cost, auc_var, batch_auc_var = dssm.combine(feature_dic_shape)

    # Test program
    test_program = fluid.default_main_program().clone(for_test=True)
    optimizer = fluid.optimizer.Adam(learning_rate=0.001)
    optimizer.minimize(avg_cost)

    print("{}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "start train reader"))
    data = reader.Dataset(feature_dict_path='./thirdparty/feature_dic.pkl')
    train_reader = paddle.v2.batch(
        paddle.v2.reader.shuffle(
            lambda: data.train_reader(train_path=cluster_train_dir), buf_size=1024),
        batch_size=BATCH_SIZE)
    print("{}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "start test reader"))
    test_reader = paddle.v2.batch(
        lambda: data.test_reader(test_path=cluster_test_dir), batch_size=BATCH_SIZE)

    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
    training_role = os.getenv("TRAINING_ROLE", "TRAINER")
    print("{} {}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "TRAINING_ROLE", training_role))
    if training_role == "PSERVER":
        place = fluid.CPUPlace()
    exe = fluid.Executor(place)
    feeder = fluid.DataFeeder(place=place, feed_list=feed_order)

    def train_loop(main_program):
        """
        train_loop
        """
        exe.run(fluid.default_startup_program())
        for pass_id in range(PASS_NUM):
            #print("pass_id" + str(pass_id))
            for batch_id, data in enumerate(train_reader()):
                print("batch_id" + str(batch_id))
                loss_tr, auc_tr, batch_auc_tr = exe.run(program=main_program, feed=feeder.feed(data),
                                                        fetch_list=[avg_cost, auc_var, batch_auc_var])
                if ((batch_id + 1) % 1) == 0:
                    print('{0} PassID {1:1}, BatchID {2:04}, Train Loss {3:2.5}, Auc {4:2.5}, Bat-Auc {5:2.5}'.
                          format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), pass_id, batch_id + 1,
                                 float(loss_tr), float(auc_tr), float(batch_auc_tr)))

                if ((batch_id + 1) % 500) == 0:
                    print('{0} PassID {1:1}, BatchID {2:04}: {3}'.
                          format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), pass_id, batch_id + 1,
                                 "start test"))
                    batch_auc_list = []
                    avg_loss_list = []
                    for tid, test_data in enumerate(test_reader()):
                        loss_t, batch_auc_t = exe.run(program=test_program,
                                                      feed=feeder.feed(test_data),
                                                      fetch_list=[avg_cost, batch_auc_var])
                        if math.isnan(float(loss_t)):
                            sys.exit("got NaN loss, training failed.")
                        batch_auc_list.append(float(batch_auc_t))
                        avg_loss_list.append(float(loss_t))
                        break

                    batch_auc_value = np.array(batch_auc_list).mean()
                    avg_loss_value = np.array(avg_loss_list).mean()
                    print('{0} PassID {1:1}, BatchID {2:04}, Test Loss {3:2.5}, Auc {4:2.5}'.
                          format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), pass_id, batch_id + 1,
                                 float(avg_loss_value), float(batch_auc_value)))

                    if int(os.getenv("PADDLE_TRAINER_ID", "0")) == 0:
                        fluid.io.save_inference_model(save_dirname + 'batch{0:04}/'.format(batch_id + 1), feed_order,
                                                      [avg_cost, auc_var], exe)
        if int(os.getenv("PADDLE_TRAINER_ID", "0")) == 0:
            fluid.io.save_inference_model(save_dirname + 'passend/', feed_order, [avg_cost, auc_var], exe)
        exe.close()
        return

    print("{}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "start train loop"))
    if is_local:
        print("start train local")
        train_loop(fluid.default_main_program())
    else:
        port = os.getenv("PADDLE_PORT", "6174")
        print("{} {}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "PADDLE_PORT", port))
        pserver_ips = os.getenv("PADDLE_PSERVERS")  # ip,ip...
        print("{} {}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "PADDLE_PSERVERS", pserver_ips))
        eplist = []
        for ip in pserver_ips.split(","):
            eplist.append(':'.join([ip, port]))
        pserver_endpoints = ",".join(eplist)  # ip:port,ip:port...
        trainers = int(os.getenv("PADDLE_TRAINERS_NUM", "0"))
        print("{} {}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "PADDLE_TRAINERS_NUM", trainers))
        current_endpoint = os.getenv("POD_IP") + ":" + port
        trainer_id = int(os.getenv("PADDLE_TRAINER_ID", "0"))
        print("{} {}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "PADDLE_TRAINER_ID", trainer_id))
        t = fluid.DistributeTranspiler()
        t.transpile(trainer_id, pservers=pserver_endpoints, trainers=trainers)
        if training_role == "PSERVER":
            pserver_prog = t.get_pserver_program(current_endpoint)
            pserver_startup = t.get_startup_program(current_endpoint, pserver_prog)
            exe.run(pserver_startup)
            exe.run(pserver_prog)
        elif training_role == "TRAINER":
            train_loop(t.get_trainer_program())


def main(use_cuda, is_local=True):
    """
    main
    """
    if use_cuda and not fluid.core.is_compiled_with_cuda():
        return

    # Directory for saving the trained model
    save_dirname = "./output/model/"
    if not os.path.isdir(save_dirname):
        os.makedirs(save_dirname)

    train(use_cuda=use_cuda, save_dirname=save_dirname, is_local=is_local)
    #infer(use_cuda, save_dirname)


if __name__ == '__main__':
    use_cuda = os.getenv("PADDLE_USE_GPU", "0") == "1"
    is_local = os.getenv("PADDLE_IS_LOCAL", "0") == "1"
    print("{} {}: {}".format(time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()), "PADDLE_IS_LOCAL", is_local))
    main(use_cuda, is_local)

**Comment by brandonsj** [comment]
reader.py

import random
import os
import pickle
import gzip


class Dataset(object):
    def __init__(self,
                 feature_dict_path='./thirdparty/feature_dic.pkl'):
        """构造函数
        """

        with open(feature_dict_path) as f:
            self.feature_dic = pickle.load(f)

    def list_file_path(self, dir_path):
        res = []
        for f in os.listdir(dir_path):
            path = os.path.join(dir_path, f)
            if os.path.isfile(path):
                res.append(path)
            elif os.path.isdir(path):
                sub_res = self.list_file_path(path)
                for item in sub_res:
                    res.append(item)
        return res

    def train_reader(self, train_path, is_filter=True):

        files = self.list_file_path(train_path)

        slot_order = [
            'a_item_type','a_category_id','a_cluster_id','a_title_segs','a_tags',
            'b_item_type','b_category_id','b_cluster_id','b_title_segs','b_tags'
        ]
        slot_num = range(2, 7) + range(8, 13)

        for fi in files:
            if os.path.isfile(fi) and fi.endswith(".gz"):
                with gzip.open(fi, 'r') as f:
                    for line in f:
                        #print(line)
                        lab, feature_slots = line.strip().split(' ', 1)
                        features = feature_slots.split(' ')
                        feature_code = [[] for i in range(10)]
                        #print("feature_code len: "+str(len(feature_code)))
                        slot = 0
                        idx = 0
                        for feature in features:
                            #print(feature)
                            feats = feature.split(',')

                            if slot in slot_num:
                                for feat in feats:
                                    #print(str(idx) + " " + str(slot))
                                    slot_name = slot_order[idx]
                                    #print(str(idx) + " " + slot_name)
                                    if feat in self.feature_dic[slot_name]:
                                        feature_code[idx].append(self.feature_dic[slot_name][feat])
                                    else:
                                        feature_code[idx].append(0)
                                idx = idx + 1

                            slot = slot + 1

                        for i in range(10):
                            if not feature_code[i]:
                                feature_code[i] = [0]
                        feature_code_new = feature_code[0:10]

                        feature_code_new.append([int(lab)])

                        #print(feature_code_new)

                        yield feature_code_new



    def test_reader(self, test_path):
        return self.train_reader(test_path, is_filter=False)


if __name__ == '__main__':
    data = Dataset('./thirdparty/feature_dic.pkl')
    train = data.train_reader('./train_data/')
    for single_sample in train:
        print(single_sample)

**Comment by brandonsj** [comment]
hybrid_dssm.py
from __future__ import print_function
import paddle.fluid as fluid

IS_SPARSE = True


def get_target_combined_features(feature_dic_shape, hid_size, init_low_bound, init_high_bound):
    emb_target_features = []

    # item_type
    a_item_type = fluid.layers.data(name='a_item_type', shape=[1], dtype='int64')
    a_item_type_emb = fluid.layers.embedding(
        input=a_item_type,
        dtype='float32',
        size=[feature_dic_shape['a_item_type'], 64],
        param_attr='a_item_type_table',
        is_sparse=IS_SPARSE)
    emb_target_features.append(a_item_type_emb)

    # category_id
    a_category_id = fluid.layers.data(name='a_category_id', shape=[1], dtype='int64')
    a_category_id_emb = fluid.layers.embedding(
        input=a_category_id,
        dtype='float32',
        size=[feature_dic_shape['a_category_id'], 64],
        param_attr='a_category_id_table',
        is_sparse=IS_SPARSE)
    emb_target_features.append(a_category_id_emb)

    # cluster_id
    a_cluster_id = fluid.layers.data(name='a_cluster_id', shape=[1], dtype='int64')
    a_cluster_id_emb = fluid.layers.embedding(
        input=a_cluster_id,
        dtype='float32',
        size=[feature_dic_shape['a_cluster_id'], 64],
        param_attr='a_cluster_id_table',
        is_sparse=IS_SPARSE)
    emb_target_features.append(a_cluster_id_emb)

    # title_segs
    a_title_segs = fluid.layers.data(name='a_title_segs', shape=[1], dtype='int64', lod_level=1)
    a_title_segs_emb = fluid.layers.embedding(
        input=a_title_segs,
        dtype='float32',
        size=[feature_dic_shape['a_title_segs'], 64],
        param_attr='a_title_segs_table',
        is_sparse=IS_SPARSE)
    a_title_segs_hidden = fluid.layers.sequence_pool(input=a_title_segs_emb, pool_type='average')
    emb_target_features.append(a_title_segs_hidden)

    # tags
    a_tags = fluid.layers.data(name='a_tags', shape=[1], dtype='int64', lod_level=1)
    a_tags_emb = fluid.layers.embedding(
        input=a_tags,
        dtype='float32',
        size=[feature_dic_shape['a_tags'], 64],
        param_attr='a_tags_table',
        is_sparse=IS_SPARSE)
    a_tags_hidden = fluid.layers.sequence_pool(input=a_tags_emb, pool_type='average')
    emb_target_features.append(a_tags_hidden)

    concat_embed = fluid.layers.concat(input=emb_target_features, axis=1)

    target_fc21 = fluid.layers.fc(input=concat_embed, size=256, act="relu")

    target_fc22 = fluid.layers.fc(input=target_fc21, size=128, act="relu")

    target_fc23 = fluid.layers.fc(input=target_fc22, size=64, act="relu")

    target_feature = target_fc23

    return target_feature


def get_source_combined_features(feature_dic_shape, init_low_bound, init_high_bound):
    emb_source_features = []

    # item_type
    b_item_type = fluid.layers.data(name='b_item_type', shape=[1], dtype='int64')
    b_item_type_emb = fluid.layers.embedding(
        input=b_item_type,
        dtype='float32',
        size=[feature_dic_shape['b_item_type'], 64],
        param_attr='b_item_type_table',
        is_sparse=IS_SPARSE)
    emb_source_features.append(b_item_type_emb)

    # category_id
    b_category_id = fluid.layers.data(name='b_category_id', shape=[1], dtype='int64')
    b_category_id_emb = fluid.layers.embedding(
        input=b_category_id,
        dtype='float32',
        size=[feature_dic_shape['b_category_id'], 64],
        param_attr='b_category_id_table',
        is_sparse=IS_SPARSE)
    emb_source_features.append(b_category_id_emb)

    # cluster_id
    b_cluster_id = fluid.layers.data(name='b_cluster_id', shape=[1], dtype='int64')
    b_cluster_id_emb = fluid.layers.embedding(
        input=b_cluster_id,
        dtype='float32',
        size=[feature_dic_shape['b_cluster_id'], 64],
        param_attr='b_cluster_id_table',
        is_sparse=IS_SPARSE)
    emb_source_features.append(b_cluster_id_emb)

    # title_segs
    b_title_segs = fluid.layers.data(name='b_title_segs', shape=[1], dtype='int64', lod_level=1)
    b_title_segs_emb = fluid.layers.embedding(
        input=b_title_segs,
        dtype='float32',
        size=[feature_dic_shape['b_title_segs'], 64],
        param_attr='b_title_segs_table',
        is_sparse=IS_SPARSE)
    b_title_segs_hidden = fluid.layers.sequence_pool(input=b_title_segs_emb, pool_type='average')
    emb_source_features.append(b_title_segs_hidden)

    # tags
    b_tags = fluid.layers.data(name='b_tags', shape=[1], dtype='int64', lod_level=1)
    b_tags_emb = fluid.layers.embedding(
        input=b_tags,
        dtype='float32',
        size=[feature_dic_shape['b_tags'], 64],
        param_attr='b_tags_table',
        is_sparse=IS_SPARSE)
    b_tags_hidden = fluid.layers.sequence_pool(input=b_tags_emb, pool_type='average')
    emb_source_features.append(b_tags_hidden)

    concat_embed = fluid.layers.concat(input=emb_source_features, axis=1)

    source_fc21 = fluid.layers.fc(input=concat_embed, size=256, act="relu")

    source_fc22 = fluid.layers.fc(input=source_fc21, size=128, act="relu")

    source_fc23 = fluid.layers.fc(input=source_fc22, size=64, act="relu")

    source_feature = source_fc23

    return source_feature

def combine(feature_dic_shape):
    hid_size = 300
    init_low_bound = -0.1
    init_high_bound = 0.1

    target_combined_features = get_target_combined_features(feature_dic_shape, hid_size, init_low_bound, init_high_bound)
    source_combined_features = get_source_combined_features(feature_dic_shape, init_low_bound, init_high_bound)
    dot_product = fluid.layers.reduce_sum(fluid.layers.elementwise_mul(target_combined_features, source_combined_features),
                                          dim=1, keep_dim=True)
    predict = fluid.layers.sigmoid(dot_product)

    binary_predict = fluid.layers.concat(input=[fluid.layers.elementwise_sub(
        fluid.layers.ceil(predict), predict), predict], axis=1)

    label1 = fluid.layers.data(name='label', shape=[1], dtype='int64')
    cost = fluid.layers.cross_entropy(input=binary_predict, label=label1)
    avg_cost = fluid.layers.mean(cost)

    auc_var, batch_auc_var, auc_states = fluid.layers.auc(input=binary_predict, label=label1)

    return [avg_cost, auc_var, batch_auc_var]

**Comment by SunGaofeng** [comment]
根据之前的讨论，补充下问题描述，
  1、本地小数据训练，可能报错，也可能不报错
  2、paddle cloud上大数据集训练，运行一段时间会报错
使用的paddle版本为fluid 1.2

**Comment by brandonsj** [comment]
This problem could be solved by using softmax_with_cross_entropy loss instead of cross_entropy.
Thank you @SunGaofeng for your help.

**Comment by Bella-Zhao** [comment]
经过排查，发现是learning_rate过大导致不收敛，pserver是正常更新的。将learning_rate调到0.002后，auc正常上升，cost下降。
